### PR TITLE
Set X-Cache header to record cache hits/misses

### PR DESCRIPTION
**Pull request by poornas**
## Description

Also set X-Cache-Lookup status to HIT if object was found in cache, but not served from cache (e.g. if Cache-Control: no-cache is set and the ETag verification failed)

## How to test this PR?
```
MINIO_CACHE_DRIVES=/tmp/cache minio gateway s3
```
`mc cat` on a pre-existing object in the cache should show X-Cache with MISS the first time, and HIT on subsequent requests.
Overwrite object on backend directly without cache and do a `mc cat` on the object - now the `X-Cache` should show `MISS` and `X-Cache-Lookup` shows `HIT` because a cache entry was found.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#328932dd-181b-4dbe-876f-cac42ebf9d7f) |
| mint-large-bucket.sh | :x: [more...](#6806e373-85bd-4c9c-9e4f-15b22c0d4c44) |
| mint-fs.sh | :x: [more...](#026ea243-e55b-4229-bbb8-699bb0d6b160) |
<a name="026ea243-e55b-4229-bbb8-699bb0d6b160"></a>
### 8794-3ca91dd/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30980
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dca3d943bb06:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 8 seconds
{"name": "awscli", "duration": 306282, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:30980/awscli-mint-test-bucket-16459""}

Executed 0 out of 13 tests successfully.
```
<a name="328932dd-181b-4dbe-876f-cac42ebf9d7f"></a>
### 8794-3ca91dd/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31184
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3a971ecd12f1:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 5171,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="6806e373-85bd-4c9c-9e4f-15b22c0d4c44"></a>
### 8794-3ca91dd/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30290
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ae99d6e11ad2:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 21 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 5 seconds
(4/13) Running mc tests ... done in 39 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 53 seconds
{
  "args": {},
  "duration": 99654,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#974142b1-2f16-423c-bc97-03cf5d7af94e) |
| mint-dist-xl.sh | :x: [more...](#8d5ae6cf-c637-4f44-b71d-892fdbc69f0f) |
<a name="974142b1-2f16-423c-bc97-03cf5d7af94e"></a>
### 8794-8b4aacb/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30643
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fff60f69dd87:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 19 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 1 minutes and 0 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 25 seconds
{
  "args": {},
  "duration": 111950,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="8d5ae6cf-c637-4f44-b71d-892fdbc69f0f"></a>
### 8794-8b4aacb/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32700
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c874d50ff41e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 8 seconds
{"name": "awscli", "duration": 306542, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:32700/awscli-mint-test-bucket-24660""}

Executed 0 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#ffbd505c-8812-4f92-8e1f-739ebce1c90d) |
| mint-large-bucket.sh | :x: [more...](#2dc0c09b-a264-4174-b3aa-85c0bc350012) |
<a name="2dc0c09b-a264-4174-b3aa-85c0bc350012"></a>
### 8794-8a9982f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30107
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 83ba02026b9d:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 15 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 5 seconds
(4/13) Running mc tests ... done in 53 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 37 seconds
{
  "args": {},
  "duration": 118104,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="ffbd505c-8812-4f92-8e1f-739ebce1c90d"></a>
### 8794-8a9982f/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30447
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9bb44040c8fd:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 15 seconds
{"name": "awscli", "duration": 312591, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:30447/awscli-mint-test-bucket-18066""}

Executed 0 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#abe999be-7d8f-4e90-9f27-6d0807ac25a4) |
| mint-large-bucket.sh | :x: [more...](#262769ad-c18f-4a28-8714-9995bc5cac2b) |
| mint-worm.sh | :x: [more...](#f1d0c228-6369-4aca-920a-9e4894b2f199) |
<a name="f1d0c228-6369-4aca-920a-9e4894b2f199"></a>
### 8794-aa4c9e3/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31826
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eb04b2153912:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-cvragava1z1jbiimu7c3npi18ypwetn4oegprmoz2slfn4h22b75nkd",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 120357,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:31826/bucket-cvragava1z1jbiimu7c3npi18ypwetn4oegprmoz2slfn4h22b75nkd: dial tcp 72.28.97.55:31826: i/o timeout",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```
<a name="262769ad-c18f-4a28-8714-9995bc5cac2b"></a>
### 8794-aa4c9e3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30837
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c00c0abb6daa:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 17 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 51 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 32 seconds
{
  "args": {},
  "duration": 106225,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="abe999be-7d8f-4e90-9f27-6d0807ac25a4"></a>
### 8794-aa4c9e3/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30731
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b68b64757dd5:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 10 seconds
{
  "name": "awscli",
  "duration": 8045,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/disk-cache.go [hunk 1]
```diff
@@ -272,8 +275,14 @@ func (c *cacheObjects) GetObjectNInfo(ctx context.Context, bucket, object string
 		default:
 		}
 	}
+
+	bkReader, bkErr := c.GetObjectNInfoFn(ctx, bucket, object, rs, h, lockType, opts)
+	// Record if cache has a hit that was invalidated by ETag verification
+	if cacheErr == nil {
+		bkReader.ObjInfo.CacheLookupStatus = CacheHit
```
Comment (kannappanr) [inline cmd/disk-cache.go:282]: What will CacheStatus be set to?

Comment (poornas) [reply cmd/disk-cache.go:282]: It defaults to MISS if not set

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**